server/src/main/java/org/elasticsearch/search/vectors/KnnScoreDocQuery.java

```diff
@@ -121,9 +130,13 @@ public long cost() {
 
                     @Override
                     public float getMaxScore(int docId) {
-                        docId += context.docBase;
+                        // NO_MORE_DOCS indicates the maximum score for all docs in this segment
+                        // Anything less than must be accounted for via the docBase.
+                        if (docId != NO_MORE_DOCS) {
+                            docId += context.docBase;
+                        }
                         float maxScore = 0;
-                        for (int idx = Math.max(0, upTo); idx < upper && docs[idx] <= docId; idx++) {
+                        for (int idx = Math.max(lower, upTo); idx < upper && docs[idx] <= docId; idx++) {
```

Comment:
Previously, this was getting the max score across ALL previous segments and the current one. This seemed weird to me. We should only return the max for the given segment, correct?

@jpountz am I correct on this?